src/kademlia/dht_storage.cpp

```diff
@@ -0,0 +1,568 @@
+					peers.erase(i++);
+				else
+					++i;
+			}
+		}
+
+		struct dht_immutable_item
+		{
+			dht_immutable_item() : value(0), num_announcers(0), size(0) {}
+			// malloced space for the actual value
+			char* value;
+			// this counts the number of IPs we have seen
+			// announcing this item, this is used to determine
+			// popularity if we reach the limit of items to store
+			bloom_filter<128> ips;
+			// the last time we heard about this
+			time_point last_seen;
+			// number of IPs in the bloom filter
+			int num_announcers;
+			// size of malloced space pointed to by value
+			int size;
+		};
+
+		struct ed25519_public_key { char bytes[item_pk_len]; };
+
+		struct dht_mutable_item : dht_immutable_item
+		{
+			char sig[item_sig_len];
+			boost::int64_t seq;
+			ed25519_public_key key;
+			char* salt;
+			int salt_size;
+		};
+
+		void touch_item(dht_immutable_item* f, address const& address)
+		{
+			f->last_seen = aux::time_now();
+
+			// maybe increase num_announcers if we haven't seen this IP before
+			sha1_hash iphash;
+			hash_address(address, iphash);
+			if (!f->ips.find(iphash))
+			{
+				f->ips.set(iphash);
+				++f->num_announcers;
+			}
+		}
+
+		// return true of the first argument is a better canidate for removal, i.e.
+		// less important to keep
+		struct immutable_item_comparator
+		{
+			immutable_item_comparator(node_id const& our_id) : m_our_id(our_id) {}
+			immutable_item_comparator(immutable_item_comparator const& c)
+				: m_our_id(c.m_our_id) {}
+
+			bool operator() (std::pair<node_id, dht_immutable_item> const& lhs
+				, std::pair<node_id, dht_immutable_item> const& rhs) const
+			{
+				int l_distance = distance_exp(lhs.first, m_our_id);
+				int r_distance = distance_exp(rhs.first, m_our_id);
+
+				// this is a score taking the popularity (number of announcers) and the
+				// fit, in terms of distance from ideal storing node, into account.
+				// each additional 5 announcers is worth one extra bit in the distance.
+				// that is, an item with 10 announcers is allowed to be twice as far
+				// from another item with 5 announcers, from our node ID. Twice as far
+				// because it gets one more bit.
+				return lhs.second.num_announcers / 5 - l_distance < rhs.second.num_announcers / 5 - r_distance;
+			}
+
+		private:
+
+			// explicitly disallow assignment, to silence msvc warning
+			immutable_item_comparator& operator=(immutable_item_comparator const&);
+
+			node_id const& m_our_id;
+		};
+	}
+
+	class dht_default_storage : public dht_storage_interface, boost::noncopyable
+	{
+	typedef std::map<node_id, torrent_entry> table_t;
+	typedef std::map<node_id, dht_immutable_item> dht_immutable_table_t;
+	typedef std::map<node_id, dht_mutable_item> dht_mutable_table_t;
+
+	public:
+
+		dht_default_storage(sha1_hash const& id, dht_settings const& settings
+			, counters& cnt);
+
+		~dht_default_storage();
+
+#ifndef TORRENT_NO_DEPRECATE
+		size_t num_torrents() const;
```

Comment:
all functions implementing the dht_storage_interface should be marked with TORRENT_OVERRIDE